TS-4265: Thread initialization update.

SolidWallOfCode opened the pull request.

PSUdaemon commented on iocore/cluster/P_ClusterCacheInternal.h:
```diff
@@ -329,14 +330,8 @@ struct CacheContinuation : public Continuation {
-    }
-    return false;
+    auto &tg = eventProcessor.thread_group[ET_CLUSTER];
+    return ET_CLUSTER && std::any_of(tg._thread, tg._thread + tg._count, [et](EThread *t) { return t == et; });
```
When is it valid that a thread in the ET_CLUSTER thread_group not be of type ET_CLUSTER?

SolidWallOfCode replied:
That's how the current code works. The type is a bit mask. But more realistically, it's cluster code, it should never be called. But I might want to put the `ET_CLUSTER` check first - if it's zero then the cluster group was never set up.

PSUdaemon commented on iocore/cache/I_Cache.h:
```diff
@@ -73,7 +73,7 @@ struct CacheProcessor : public Processor {
   }
 
-  virtual int start(int n_cache_threads = 0, size_t stacksize = DEFAULT_STACKSIZE);
+  virtual int start(int n_cache_threads = 0, size_t stacksize = DEFAULT_STACK_SIZE) override;
```
How come the name of the constant is changed to add an underscore but the variable name is not?

SolidWallOfCode replied:
Heh. I'll look at that.

PSUdaemon commented on iocore/eventsystem/I_EventProcessor.h:
```diff
@@ -112,16 +118,18 @@ class EventProcessor : public Processor
+
+  /// Convenience overload.
+  /// This registers @a name as an event type using @c registerEventType and then calls the real @c spawn_event_threads
+  EventType spawn_event_threads(const char *name, int n_thread, size_t stacksize = DEFAULT_STACK_SIZE);
```
This feels icky to me. 🚲

SolidWallOfCode replied:
It's handy in several places.

PSUdaemon commented on doc/developer-guide/architecture/threads-and-events.en.rst:
```diff
@@ -0,0 +1,135 @@
+   eventProcessor.spawn_event_threads(ET_GROUP, n_group_threads, GROUP_STACK_SIZE);
+
+
+The function :code:`Group_Thread_Init` can be replaced with a continuation if that's more convenient. One advantage of a continuation is additional data (via :arg:`cookie`) can be provide during thread initialization.
```
Is `edata` being replaced with `cookie`? Seems like a name collision to use `cookie`.

SolidWallOfCode replied:
I was mixed about that. It's called a `cookie` in the `Event` structure, which is where the argument name comes from.

PSUdaemon commented on iocore/eventsystem/UnixEThread.cc:
```diff
@@ -41,27 +41,12 @@ struct AIOCallback;
-    signal_hook(nullptr),
-    tt(att),
-    server_session_pool(nullptr)
+EThread::EThread(ThreadType att, int anid) : id(anid), tt(att), server_session_pool(NULL)
```
Why changing `nullptr` to `NULL`?

SolidWallOfCode replied:
Because this was done initially before the Great Null Shift and I missed this during the merge.

PSUdaemon commented on iocore/eventsystem/UnixEventProcessor.cc:
```diff
@@ -32,226 +32,216 @@
+  /// Set the affinity for the current thread.
+  int set_affinity(int, Event *);
+
+#if defined(HAVE_HWLOC_OBJ_PU)
```
Shouldn't this be `#if defined(TS_USE_HWLOC)`?

PSUdaemon commented on iocore/eventsystem/UnixEventProcessor.cc:
```diff
@@ -32,226 +32,216 @@
+    Debug("iocore_thread", "EThread: %d %s: %d CPU Mask: %s\n", i, _name, obj->logical_index, cpu_mask);
 #else
-      Debug("iocore_thread", "EThread: %d %s: %d", i, obj_name, obj->logical_index);
+    Debug("iocore_thread", "EThread: %d %s: %d\n", i, _name, obj->logical_index);
```
Why adding newline char here?

zwoop committed:
Bump config delete timeout to 5min

scw00 committed:
obj_found is never be true

danobi committed:
Clean up BaseLogFile

Let's use some C++11 features!

persiaAziz-zz committed:
remap test for secure communication with origin server

bryancall committed:
coverity 1196479: Uninitialized pointer field

PSUdaemon committed:
Coverity CID #1241990 Resource leak

jpeach committed:
Remove ending period from error messages.

Standardize error messages by removing the terminating period.

sfeltner-godaddy committed:
Coverity 1021693: Uninitialized pointer field

vmamidi committed:
remove the configuration to enable or disable safe requests retry

bryancall committed:
Updated clang-tidy default config to be the option that have been committed

bryancall committed:
coverity 1224618: Uninitialized scalar field

taoyunxing committed:
Output the responding url when "Warning: Document truncated" occurs for a valid
request.

PSUdaemon committed:
Coverity CID #1367529 Copy into fixed size buffer

PSUdaemon committed:
Coverity CID #1022089 Copy into fixed size buffer

PSUdaemon committed:
Coverity CID #1022090 Copy into fixed size buffer

PSUdaemon committed:
Coverity CID #1241990 Resource leak

persiaAziz-zz committed:
Coverity_1022180_1022181: ka_action can't be KA_UNKNOWN in the switch statement

Commit:
proxy.config.diags.debug.enabled is reloadable

zwoop committed:
Various Coverity fixes for logstats

	Coverity 1199997
	Coverity 1200000
	Coverity 1200003
	Coverity 1200012

zwoop committed:
Adds a coverity model file, and excludes a stat() check

This also undoes a previous change, that hopefully the new model rule
will take care of (instead of nesting the if/else).

scw00 committed:
coredump with server_session

persiaAziz committed:
Coverity_1021917: closet_match can not be null at that point

PSUdaemon committed:
Coverity CID #1022090 Copy into fixed size buffer

zwoop committed:
Add a couple of CI build scripts, to source from Jenkins

Commit:
update epel version.

persiaAziz-zz committed:
Issue_1943: get rid of ticket_key_name tag from ssl_multicert

zwoop committed:
More cleanup of the CI scripts

danobi committed:
Delete .indent.pro

This is for the GNU indent tool, which has been superseded by clang-format.

danobi committed:
Delete .gitmodules

The file is empty so I don't think we need it.

calavera committed:
Fix TSMilestonesType documentation.

The name of this type is misspelled in the documentation.

Signed-off-by: David Calavera <david.calavera@gmail.com>

zwoop committed:
Use proper make wrapper, and parallelize

SolidWallOfCode committed:
Doc: Fixup milestones documentation.

scw00 committed:
coredump_handleEvent fix

pbchou committed:
Plugin ts_lua - Incorrect mapping of transaction configuration parameter.

TS_LUA_CONFIG_HTTP_NEGATIVE_CACHING_LIFETIME was mapped to the wrong
TS_CONFIG_... value. This patch corrects the mapping thus allowing
the negative caching lifetime to be changed via the Lua plugin.

a-canary committed:
Deprecated VConnection do_io

Replaced all calls to do_io(VIO::ENUM, ...) with matching do_io_*(...).

Commit:
ignore .vscode/

scw00 committed:
Log the Lua error message when evaluating metrics fails.

oknet committed:
spawn_thread with atomic operation

The eventProcessor.spawn_thread() is used to create a DEDICATED EThread.
The n_dthreads is total number of DEDICATED EThreads. When a new
DEDICATED EThread created, it makes the `all_dthreads[n_dthreads]' point
to the EThread object then do `n_dthreads++'

It is not likes the spawn_event_threads() that only called from main()
before any EThread runs. It is can be called from main() and/or any
EThread.

Therefore there is a race on accessing `all_dthreads[n_dthreads]' and
`n_dthreads++'.

zwoop committed:
Removes the tsqa tests, to avoid confusion. Use autest.

bryancall committed:
Fixed problem with url_len sometimes not being initialized

SolidWallOfCode committed:
Doc: Fix up some TS SSL API documentation warnings.

SolidWallOfCode committed:
TS-4265: Thread initialization update.

SolidWallOfCode committed:
TS-4265: Ready for testing.

SolidWallOfCode committed:
TS-4265: Fix start up race condition.

SolidWallOfCode committed:
TS-4265: The Phil Edit.
